pkg/volume/volume.go

```diff
@@ -65,6 +66,12 @@ type Recycler interface {
 	Recycle() error
 }
 
+// Create adds a new resource in the storage provider and creates a PersistentVolume for the new resource.
+// Calls to Create should block until complete.
+type Creater interface {
+	Create() (*api.PersistentVolume, error)
```

Comment:
> Thinking through that a bit -- so we will wind up with a provisioner for GCEPD that gets the cloud provider from the VolumeHost interface, the makes a downcast to GCECloudProvider, that uses some special function of that type to create a new PD.

Correct - it's just a matter of where the code lives.

> That seems simple enough, but it's not going to be as cut and dry for other volume types. NFS, for example:
>
> You want a provisioner that knows how to provision EFS volumes that are used as NFS volumes
> You want a provisioner that knows how to provision using the netapp ONTAP API
> What happens when you have a cluster with an aws cloud provider and you want to provision some NFS volumes via EFS and some via ONTAP against a machine running the ONTAP server in AWS?

The NetApp NFS provisioner is a totally different beast than the GCE
provisioner.    I do not think we should have a single provisioner
that knows how to provision every volume type.

> Maybe we solve (3) by differentiating distinct volume types for differently provisioned volumes of the same type, but that doesn't sit quite right with me. I guess the question is whether there is really utility for an encompassing 'type' for each permutation of (client-side-concerns,provisioning-concerns) versus decoupling them.
>
> —
> Reply to this email directly or view it on GitHub.